rust/arrow/src/compute/kernels/take.rs

```diff
@@ -166,42 +166,124 @@ fn take_primitive<T>(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRe
 where
     T: ArrowPrimitiveType,
 {
-    let mut builder = PrimitiveBuilder::<T>::new(indices.len());
-    let a = values.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
-    for i in 0..indices.len() {
-        if indices.is_null(i) {
-            // populate with null if index is null
-            builder.append_null()?;
-        } else {
-            // get index value to use in looking up the value from `values`
-            let ix = indices.value(i) as usize;
-            if a.is_valid(ix) {
-                builder.append_value(a.value(ix))?;
-            } else {
-                builder.append_null()?;
+    let data_len = indices.len();
+
+    let array = values.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+
+    let num_bytes = bit_util::ceil(data_len, 8);
+    let mut null_buf = MutableBuffer::new(num_bytes).with_bitset(num_bytes, false);
```

Comment:
About 13% improvement on my computer:

```
take str 512            time:   [6.6736 us 6.6928 us 6.7177 us]                          
                        change: [-15.907% -12.819% -10.196%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe

take str 1024           time:   [11.579 us 11.597 us 11.614 us]                           
                        change: [-14.361% -12.760% -11.299%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  5 (5.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe
```

I do not know how to address the `data_len % 8 != 0`, though: I can't find the APIs to handle bits like that.

Reply:
Is the change worth it? We can also try it on bool and numeric arrays.

If you have 35 values in take, and num_bytes = 5, you can create a range `((35 - 1)..(num_bytes * 8)).for_each(|i| unset_bit);`I'll play with godbolt.org to see if the compiler is smart enough to use BMI instructions for the above range.

Reply:
This might also be a non-issue as tests pass without changing the bits, and fail when changing them. We'll know if it's a problem if integration tests fail, so maybe we can address it later.